illumos-utils/src/host/executor.rs

```diff
@@ -0,0 +1,532 @@
+use std::io::{Read, Write};
+use std::process::{Command, Stdio};
+use std::str::from_utf8;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+pub fn command_to_string(command: &std::process::Command) -> String {
+    command
+        .get_args()
+        .map(|s| s.to_string_lossy().into())
+        .collect::<Vec<String>>()
+        .join(" ")
+}
+
+fn to_space_separated_string<T, I>(iter: T) -> String
+where
+    T: IntoIterator<Item = I>,
+    I: std::fmt::Debug,
+{
+    Itertools::intersperse(
+        iter.into_iter().map(|arg| format!("{arg:?}")),
+        " ".into(),
+    )
+    .collect::<String>()
+}
+
+fn log_input(log: &Logger, id: u64, command: &Command) {
+    info!(log, "{id} - Running Command: [{}]", Input::from(command),);
+    debug!(
+        log,
+        "{id} - Environment: [{}]",
+        to_space_separated_string(command.get_envs()),
+    )
+}
+
+fn log_output(log: &Logger, id: u64, output: &Output) {
+    info!(
+        log,
+        "{id} - {} (status code: {})",
+        if output.status.success() { "OK" } else { "ERROR" },
+        output
+            .status
+            .code()
+            .map(|c| c.to_string())
+            .unwrap_or_else(|| "none".to_string()),
+    );
+    if !output.stdout.is_empty() {
+        debug!(
+            log,
+            "{id} - stdout: {}",
+            from_utf8(&output.stdout).unwrap_or("<Not valid UTF-8>"),
+        );
+    }
+    if !output.stderr.is_empty() {
+        debug!(
+            log,
+            "{id} - stderr: {}",
+            from_utf8(&output.stderr).unwrap_or("<Not valid UTF-8>"),
+        );
+    }
+}
+
+/// Describes the commonly-used "safe-to-reference" type describing the
+/// Executor as a trait object.
+pub type BoxedExecutor = Arc<dyn Executor>;
+
+/// Describes an "executor", which can run [Command]s and return a response.
+///
+/// - In production, this is usually simply a [HostExecutor].
+/// - Under test, this can be customized, and a [FakeExecutor] may be used.
+#[async_trait]
+pub trait Executor: Send + Sync {
+    /// Executes a task, waiting for it to complete, and returning output.
+    async fn execute_async(
+        &self,
+        command: &mut tokio::process::Command,
+    ) -> Result<Output, ExecutionError>;
+
+    /// Executes a task, waiting for it to complete, and returning output.
+    fn execute(&self, command: &mut Command) -> Result<Output, ExecutionError>;
+
+    /// Spawns a task, without waiting for it to complete.
+    fn spawn(
+        &self,
+        command: &mut Command,
+    ) -> Result<BoxedChild, ExecutionError>;
+}
+
+/// Handler called when spawning a fake child process
+pub type SpawnFn = dyn FnMut(&mut FakeChild) + Send + Sync;
+pub type BoxedSpawnFn = Box<SpawnFn>;
+
+/// Handler called when awaiting a fake child process
+pub type WaitFn = dyn FnMut(&mut FakeChild) -> Output + Send + Sync;
+pub type BoxedWaitFn = Box<WaitFn>;
+
+pub(crate) struct FakeExecutorInner {
```

Comment:
I don't think they can be if other crates want to use them, right? The crate being tested is compiled with `cfg(test)`, but its dependencies aren't. I suggested moving the fake-support stuff into a module of its own, but maybe it should be a _crate_ of its own? That way we could add it to `dev-dependencies` from any crate that wanted to use it.

Reply:
Ah, you're right, I completely forgot about the crate boundary there. Mea culpa. I think a separate crate makes sense, since (at least in my head) there wouldn't really be any code shared between the fake "host callouts" implementation and the real one (except for the trait definition that encapsulates the interface to both).

If they did have to share code for some reason, perhaps we could just have a separate feature guard that enables compilation of the fake? Not having thought about it too deeply, I think either method would get us toward a world where we can build "production" binaries such that (a) we're statically assured we didn't somehow manage refer to the fake, and (b) we don't have to spend any time compiling a feature-filled test fake we're not using :)

Reply:
See https://github.com/oxidecomputer/omicron/pull/3442#issuecomment-1687179358 below -- the fakes exist in the `tokamak` crate, the "real" implementation exists in the `protostar` crate, and the interfaces now exist in the `fusion` crate.